### ISSUE TITLE
Desktop treemacs helper buffer not closing

### ISSUE THREAD

**Issue by technician77** [Alexander-Miller/treemacs#101]
When I start emacs I got this desktop treemacs helper buffer. It says text below. But I can't call any function that is called treemacs-restore or similar. I assume something is wrong here.

```
This buffer only exists so that desktop mode will load treemacs.
It will be deleted by `treemacs--restore'.This buffer only exists so that desktop mode will load treemacs.
It will be deleted by `treemacs--restore'.
```

**Comment by Alexander-Miller** [comment]
Known edge case, caused by e.g. no treemacs-buffer being revived because you already opened one for that frame before lazy restore went off. Killing that buffer yourself is completely safe.

Fix will be pushed later today.

**Comment by Alexander-Miller** [comment]
Fix is a go, desktop helper should now disappear properly.

**Comment by technician77** [comment]
First, you should know that I'm currently on Emacs 26.0.90. I've observed the following:
- Desktop save mode seems to cause the problems
- Scratch buffer is broken when treemacs is installed. It now comes up in fundamental mode without a message. Gets back to normal when treemacs is removed
- Even with treemacs disabled in use-package it is started, as said above, Desktop-save mode seems to spawn it.
- manual save of the desktop causes the treemacs desktop buffer to spawn. That is really wreird.

**Comment by Alexander-Miller** [comment]
> Scratch buffer is broken when treemacs is installed. It now comes up in fundamental mode without a message. Gets back to normal when treemacs is removed

I think that's because I return the scratch buffer in a spot to make desktop-mode think all is well when I shouldn't. Spacemacs probably makes it look correct for me anyway. Let's test this: 
Find treemacs-persist.el, go to line 139, replace the `get-buffer-create` part with just a call to `(current-buffer)`, run `emacs-lisp-byte-compile` and restart.

> Even with treemacs disabled in use-package it is started, as said above, Desktop-save mode seems to spawn it.

That looks like desktop-mode is working is intended. It sees a buffer in treemacs-mode, so tries to load treemacs.el. If treemacs is still in your `load-path` it will be found and loaded and restored.

> manual save of the desktop causes the treemacs desktop buffer to spawn. That is really wreird.

That's because I haven't considered the manual option. Long story short: For session persistence I map treemacs buffers to frames via frame parameters. Desktop restores first buffers, then frame params. So when persisting I save my state, then kill all treemacs buffers and leave only the desktop helper, so desktop mode will load treemacs. With desktop-mode also having a lazy load option having just a single treemacs-mode buffer makes it much easier to create all the workarounds I need to make sure treemacs buffers are restored and frame-mapped correctly.

Looks like I need to rethink my current approach.

**Comment by ylluminarious** [comment]
I've been having quite similar issues as well.

**Comment by technician77** [comment]
I replaced the expression as suggested and ran the byte-compile but no change regarding scratch.

**Comment by Alexander-Miller** [comment]
Just pushed a new version. The desktop helper will now only spawn on emacs shutdown. Let me know if it works.

**Comment by technician77** [comment]
Tried treemacs-20171118.229. Saving the desktop doesn't respawn the desktop helper buffer, but restarting does. Scratch is still broken. Tried (setq treemacs-never-persist t) but no change. Even when parameter is set treemacs-persist is still written to disk. Is this normal?

**Comment by Alexander-Miller** [comment]
> Even when parameter is set treemacs-persist is still written to disk. Is this normal?

No, that's a bug. I've forgotten about never-persist between all the refactoring.

> Saving the desktop doesn't respawn the desktop helper buffer, but restarting does.

The helper should be short-lived to the point that you shouldn't even see it. If it's restored eagerly it will be killed in the `desktop-after-read-hook`. If it's restored lazily it will be killed right before treemacs starts reviving its buffers. 

I've now pushed a patch to take never-persist into account. The persist process will now delete treemacs' data instead of writing it when never-persist is t and the desktop is saved. That'll prevent treemacs being loaded on the next start.

I suggest you try again with a clean restart: Delete treemacs and its persist data and try again. The persist data is stored in `$user-emacs-directory/cache/treemacs-persist`. Better delete your saved desktop data as well. It's possible that some broken version of scratch got stuck there from a previous treemacs version At least that's the only explanation that makes sense to me, since treemacs's entire code base now has zero interactions with the scratch buffer.

**Comment by technician77** [comment]
That was weird. I used 20171118.601 and on the first start it showed the same behaviour, showing Desktop Helper buffer and Scratch in Fundamental. Despite I deleted it, removed cache and deleted the desktop save file. Additionally I installed manually, but didn't M-x and put (setq treemacs-never-persist t) in the :init section of use-package.

Strangely on the second start of emacs, everything worked as expected. No desktop helper buffer anymore and scratch came up normally. I opened treemacs and things stayed that way even after a restart. I put (setq treemacs-never-persist t) back in the :config section since I think it doesn't make a difference.

I did revert back to see what happens if  treemacs-never-persist is set to nil. It did revert back to the broken behaviour. Setting it back to t made it work again as expected.

**Comment by Alexander-Miller** [comment]
I still cannot reproduce this, so we need to take a different approach. I need to get a picture of what's happening on your system. To that end I have uploaded the `persist` branch, which you should use for your local installation. It will be logging treemacs' persist and restore process to ~/Documents/TREEMACS-LOG.
Install the branch, start with a clean desktop, run treemacs persistence and restoration, and then show me what the content of the log file looks like.

On my system saving and then correctly restoring a single treemacs buffer looks like this:
```
[Treemacs] Call to persist helper creation
[Treemacs] Creating persist helper *Desktop Treemacs Helper*
[Treemacs] Call to desktop persist advice
[Treemacs] Call to persist
[Treemacs] Writing persist state (((scope-id . 1) (root . /home/a/Documents/git/treemacs) (point . /home/a/Documents/git/treemacs/treemacs-persist.el)) ((scope-id . 1) (root . /home/a/Documents/git/treemacs) (point . /home/a/Documents/git/treemacs/treemacs-persist.el)))
[Treemacs] Call to desktop handler
[Treemacs] Call into restore
[Treemacs] Never persist = nil
[Treemacs] Id frame params not yet restored = nil
[Treemacs] Starting timer
[Treemacs] Complete call to restore, returning buffer treemacs-persist.el
[Treemacs] In timer
[Treemacs] Killing desktop helper *Desktop Treemacs Helper*
[Treemacs] Found persist data (((scope-id . 1) (root . /home/a/Documents/git/treemacs) (point . /home/a/Documentsinstall/git/treemacs/treemacs-persist.el)) ((scope-id . 1) (root . /home/a/Documents/git/treemacs) (point . /home/a/Documents/git/treemacs/treemacs-persist.el)))
[Treemacs] Restoring Buffer  *Treemacs-Framebuffer-1* to Frame #<frame emacs-25.3@a-laptop 0x11f5c30>
```

Looking at this output I just noticed that the written state seems doubled. This is a bug, but should not have any side effects since a buffer won't be restored if its frame already has a treemacs buffer.

**Comment by Alexander-Miller** [comment]
Double entry bug is fixed in both persist and master.

**Comment by technician77** [comment]
I'll try that asap. Could it be that you reverted the behaviour back so that (setq treemacs-never-persist t) is ignored again on the main branch?

**Comment by Alexander-Miller** [comment]
No, the code is all as it should be. The persist function will clearly delete your saved state when never-persist is t:
```
  (if treemacs-never-persist
      (when (f-exists? treemacs--persist-file)
        (f-delete treemacs--persist-file))
    (-let [persist-dir (f-dirname treemacs--persist-file)] ....
```

I cannot say why your system is behaving the way it does. The logs will show more. You should do multiple runs, one with never-persist set to t, and one without.
The logging is a simple ad-hoc implementation and will just keep appending to the file, so the log file should be deleted between every run.

**Comment by technician77** [comment]
So I did some testing using treemacs-persist:

- When using treemacs-never persist is set to nil, then scratch is broken.
- When it is set to t it works as intended ON THIS BRANCH
- When using the main branch scratch is broken in both states
- Main branch got broken after you created the persist branch (coincidence?)
--------------------------------------------------------------------------------------------------------------------
- Config
```
  (use-package treemacs
    :ensure nil
    :bind
    ("<f12>" . treemacs-toggle)
    :config
    ;; (setq treemacs-never-persist t)
    (setq treemacs-follow-mode t)
    (setq treemacs-filewatch-mode t)
    (setq treemacs-width 100))
```
	
- What was done	
desktopfile+cache non-existant, started emacs, scratch o.k., treemacs called via hotkey, then closed, desktop was saved

```
[Treemacs] Call to persist helper creation
[Treemacs] Creating persist helper *Desktop Treemacs Helper*
[Treemacs] Call to desktop persist advice
[Treemacs] Call to persist
[Treemacs] Creating persist dir z:/home/myusername/.emacs.d/.cache
[Treemacs] Writing persist state (((scope-id . 1) (root . z:/Some-Path) (point . z:/Some-Path/2002)))
```
--------------------------------------------------------------------------------------------------------------------
- Config
```
  (use-package treemacs
    :ensure nil
    :bind
    ("<f12>" . treemacs-toggle)
    :config
    ;; (setq treemacs-never-persist t)
    (setq treemacs-follow-mode t)
    (setq treemacs-filewatch-mode t)
    (setq treemacs-width 100))

```
- What was done
restarted emacs with desktop and cache just saved, scratch broken, treemacs called via hotkey, then closed, 

```
[Treemacs] Call to desktop handler
[Treemacs] Call into restore
[Treemacs] Never persist = nil
[Treemacs] Id frame params not yet restored = t
[Treemacs] Complete call to restore, returning buffer *scratch*
[Treemacs] Call into restore
[Treemacs] Never persist = nil
[Treemacs] Id frame params not yet restored = t
[Treemacs] Call to persist helper creation
[Treemacs] Creating persist helper *Desktop Treemacs Helper*
[Treemacs] Call to desktop persist advice
[Treemacs] Call to persist
[Treemacs] Writing persist state (((scope-id . 1) (root . z:/home/myusername) (point . z:/home/myusername/.config)))
```
--------------------------------------------------------------------------------------------------------------------
- Config
```
  (use-package treemacs
    :ensure nil
    :bind
    ("<f12>" . treemacs-toggle)
    :config
    (setq treemacs-never-persist t)
    (setq treemacs-follow-mode t)
    (setq treemacs-filewatch-mode t)
    (setq treemacs-width 100))
```
	
- What was done
desktopfile+cache non-existant, started emacs, scratch o.k., treemacs called via hotkey, then closed, desktop was saved

```
[Treemacs] Call to persist helper creation
[Treemacs] Call to desktop persist advice
[Treemacs] Call to persist

```
--------------------------------------------------------------------------------------------------------------------
- Config
```
  (use-package treemacs
    :ensure nil
    :bind
    ("<f12>" . treemacs-toggle)
    :config
    (setq treemacs-never-persist t)
    (setq treemacs-follow-mode t)
    (setq treemacs-filewatch-mode t)
    (setq treemacs-width 100))
```
	
- What was done
restarted emacs with desktop and cache just saved, scratch o.k., treemacs called via hotkey, then closed

```
[Treemacs] Call to persist helper creation
[Treemacs] Call to desktop persist advice
[Treemacs] Call to persist

```

**Comment by Alexander-Miller** [comment]
I see 2 issues here:

 * I give desktop-mode the current buffer since it needs a buffer object to count restoration a success. For you that's usually the scratch buffer, which gets broken in the process somehow. I've  now changed this to return the desktop helper instead, creating it if necessary (for me it doesn't even live during desktop restoration).

 * `Id frame params not yet restored = t` is not supposed to stick around permanently. Desktop mode is supposed to restore your frame parameters eventually. That would then allow treemacs to know which buffer goes with which frame. Desktop mode is suppose to restore frame parameters, no matter whether treemacs restoration works or not. 

For the latter let's try the following. Start emacs and treemacs and run below script to confirm that you have a frame mapping to a number. Restart emacs and wait for desktop mode to do its thing, then rerun the script. The mapping it prints should look the same as before.

```
(dolist (frame (frame-list))
  (message "%s -> %s" frame (frame-parameter frame 'treemacs-id)))
```

**Comment by technician77** [comment]
Things improved for me. Now scratch buffer always works. If I change never persist to t or not, scratch remains working. The desktop helper buffer is gone too, it poped up once between (re)installation an now but that's o.k. I just hope it stays that way.

```
#<frame  *Minibuf-1*    0000000009968c50> -> 1
#<frame F1 0000000400616480> -> nil

#<frame  *Minibuf-1*    0000000009711340> -> nil
#<frame F1 0000000400616480> -> nil

#<frame  *Minibuf-1*    0000000009711740> -> 1
#<frame F1 0000000400616480> -> nil

#<frame  *Minibuf-1*    0000000009979df0> -> 1
#<frame F1 0000000400616480> -> nil

#<frame  *Minibuf-1*    00000000099730e0> -> 1
#<frame F1 0000000400616480> -> nil
```

**Comment by Alexander-Miller** [comment]
Persistence is working correctly for you then? Or is there still something that's missing?

> *Minibuf-1*

It's a bit odd that you have a frame called Minibuf, especially since it's the frame treemacs is attached to.

@ylluminarious How about you? Is persistence working as intended with the newest changes?

**Comment by technician77** [comment]
Yes, it does work. Although that the bugs are gone is more important for me.

I don't know about Minibuf-1. The lisp was evaluated with a split window having treemacs on the left and scratch on the right. Well, Emacs 26 is not yet released and on the other hand there is a multitude of other packages installed and the init is large as well, maybe some code is bending it that way.

Thanks alot for taking care of the issue,
Kind regards mnemonic

**Comment by ylluminarious** [comment]
@Alexander-Miller @technician77 After updating `treemacs`, the issue seems to be gone for me. I haven't followed the rest of the discussion between the two of you very closely, but I can say that things seem fine for me.

**Comment by Alexander-Miller** [comment]
TL;DR: don't give desktop-mode your scratch buffer.

This bit of trouble seems to be over, so I'll close up here.

**Comment by technician77** [comment]
Well, I have to come back to this. The scratch buffer issue was solved back than and has never appeared again. The other problem, the `*Desktop Treemacs Helper*` buffer remained. During troubleshooting last year the problem was at some stage gone, but returned on one of the last versions during the troubleshooting last year. I thought to myself, well the major problem is gone and maybe this will be gone at some point along the updates. Sadly even with emacs 26.1 RC1 and the latest treemacs the problem remains. If I kill the helper buffer it returns on restart. I might bury the buffer or something but if you got time you might have a look.
I run emacs 26.1 RC1 on windows. A treemacs-persit file does not exist. My configuration is this:
  (use-package treemacs
    :bind
    ("<f12> t" . treemacs-toggle)
    :config
    (setq treemacs-never-persist t)
    (setq treemacs-follow-mode t)
    (setq treemacs-filewatch-mode t)
    (setq treemacs-width 100))

**Comment by Alexander-Miller** [comment]
The current system will be gone in the next major update. I've decided to remove the integration with desktop-save-mode due to exactly this kind of instability. The replacement is much simpler - treemacs wont be reinitialized automatically, but will simply remember what it used to look like the last time emacs was shut down.

Desktop-save-mode integration mayor may not come back. Depends on whether I find a better way to make it work than it does now,
